install-deps.sh

```diff
@@ -411,6 +417,11 @@ else
         sed "/$IGNORE_YUM_BUILDEP_ERRORS/d" $DIR/yum-builddep.out | grep -qi "error:" && exit 1
         ;;
     opensuse*|suse|sles)
+        if [ "$ID" = "opensuse-leap" -o "$ID" = "sles" ]; then
+            if version_lt $VERSION_ID 15; then
+                python_interpreters+=" python2.7"
+            fi
+        fi
```

Comment:
Please drop this. Leap 15.* (SLE-15-*) are py3 distros.

Reply:
i don't understand your reasoning here. what i want is something like (in python pseudo code)
```py
if id in ('opensuse-leap', 'sles):
  if version < 15:
    python_interpreters.append('python2.7')
```
does this make more sense?

Reply:
Can you help me understand why this code block is needed? The SUSE version before 15 was 12. Now, 12 was a py2 distro, but the last Ceph release that ran on 12 was luminous.

Reply:
sure. it depends on how we should send the message that we don't support some distros. if a developer running, for example, xenial or any openSUSE other than Leap or Tumbleweed, shall we just send a message that we don't support this distro? if we agree on this, i am fine with repurposing this code block and bailing out this script with an error message.

Reply:
Regardless of which distro one is using, master/Octopus only runs on more recent versions of the distro. If someone tries to run `./install-deps.sh` on a very early version of CentOS, RHEL, Ubuntu, or Debian, does the script currently bail out? Is adding that functionality relevant to this PR?

(BTW I think we should be careful with the word "support"....)

Reply:
> If someone tries to run `./install-deps.sh` on a very early version of CentOS, RHEL, Ubuntu, or Debian, does the script currently bail out?

i am not interested in keeping the existing behavior in this perspective. what i am interested in is to help the developers like Thomas in #31434. as long as it helps them by showing an error message, i think it's nice to have.

> (BTW I think we should be careful with the word "support"....)

actually, i am not sure what do you mean by "run". to my best understanding, we've dropped the support of xenial. not sure about OpenSUSE or SLES though.

Reply:
Thomas in #31434 was running on Tumbleweed, so this proposed code would not help him.

Does install-deps.sh currently abort with an error message when someone runs it on Xenial or before? I have not tried it. If so, I am in favor of adding that behavior to install-deps.sh for any distros where it is missing.

Upstream does not "support" anything, that's why I recommend to avoid that word in upstream contexts. It confuses people, because different people have different expectations of "support".

Reply:
If `install-deps.sh` does not help uninformed developers on Ubuntu, CentOS etc. by providing an error message when the OS version is not new enough, then I don't understand why such an error message should be added for SUSE. Either add the error message for all, or for none.

Reply:
please note, i am talking about the expected behavior instead of the existing behavior. currently, `install-deps.sh` does not error out because of the some distro it does not support. also, "showing an error message" is not specific to OpenSUSE or SLES. we started this discussion in this very code block, but this discussion is not limited this code block.

@smithfarm  i don't want to argue what is "supported" or what do we support. i am referencing to https://docs.ceph.com/docs/master/start/os-recommendations/#platforms . i am sorry if this confuses any of us.

Reply:
Yes, I understood what you were talking about. Here's what I don't understand, though: what I hear you saying is you'd like to implement a new feature (error out on older versions of the OS) in this PR, but only for SUSE?

Reply:
As long as the "error out on older versions of OS" feature is implemented for all OSes equally, then I think it's good thing to have, but would rather see it in a different PR since it's not related to the issue this PR is addressing.

Reply:
> so this proposed code would not help him.

i think you missed the point of this change. the idea is to avoid adding python2.7 to "python_interpreters" for "new" distros like OpenSUSE Tumbleweed, and hence avoid building wheels for python2.7. the issue Thomas ran into is that he didn't have python2.7, but install-deps.sh tried to build wheels using python2.7. i think i explained this already in the commit message.

Reply:
I talked to Thomas, and the issue he had was different. He's running Tumbleweed and the problem appeared because he had both py2 and py3 in the same system. Without py2 installed, "make check" runs fine on Tumbleweed.

Reply:
In any case, this `if` block can be dropped. It does not help to address Thomas's issue.

Reply:
> what I hear you saying is you'd like to implement a new feature (error out on older versions of the OS) in this PR, but only for SUSE?

how come you have this understanding? i never want to make SUSE an exception, that exactly why i added the code block you want to remove.

Reply:
well, in the case of the code block you added, SUSE *is* the exception. We definitely don't want to test py2 on any recent SUSE distro.

Reply:
Apparently, there was a misunderstanding that Thomas was using an outdated distro and that's why he had a problem?

Reply:
I guess what I'm trying to say is this: this PR will probably fix Thomas's issue regardless of whether this code block stays or is dropped. I asked for it to be dropped because (1) it doesn't solve any problem that any actual developer is having and (2) it introduces py2 code into install-deps.sh in master, where we are trying to get rid of py2.

Reply:
> We definitely don't want to test py2 on any recent SUSE distro.

see https://github.com/ceph/ceph/pull/31525#discussion_r344936460. i don't think < 15 is recent.

> it doesn't solve any problem that any actual developer is having

i don't have an assumption of actual developer. but i trust yours.

> it introduces py2 code into install-deps.sh in master

again, i am not introducing py2 code. it's already there. i am trying to avoid adding python2.7 for python3 distros.

but anyhow, i am dropping this piece of code.